Jenkinsfile

```diff
@@ -0,0 +1,165 @@
+                 docker push ${repo} && \\
+                 docker rmi ${repo} && \\
+                 docker logout'
+        """
+    }
+}
+
+// Containers may not exit immediately if a build is aborted through Jenkins. Explicitly kill the container to
+// make sure it stops. Also remove the container for good measure. Just because I'm paranoid, it doesn't mean
+// that they're not out to get me.
+def removeContainer(containerName) {
+    sh returnStatus: true, script: "docker kill ${containerName}"
+    sh returnStatus: true, script: "docker rm ${containerName}"
+}
+
+// Ensure that a Docker image is removed.
+def removeImage(image) {
+    sh returnStatus: true, script: "docker rmi ${image}"
+}
+
+// Records build information in the associated Jira issue if possible.
+def recordBuild(label) {
+    step(
+        [
+            $class: 'hudson.plugins.jira.JiraIssueUpdater',
+            issueSelector: [$class: 'hudson.plugins.jira.selector.DefaultIssueSelector'],
+            scm: scm,
+            labels: [label]
+        ]
+    )
+}
+
+// Configure this job to clean up old builds.
+properties(
+    [
+        buildDiscarder(
+            logRotator(
+                artifactDaysToKeepStr: '',
+                artifactNumToKeepStr: '',
+                daysToKeepStr: '7',
+                numToKeepStr: '5'
+            )
+        )
+    ]
+)
+
+// The packages we want to test.
+packagesToTest = "github.com/cyverse-de/dataone-indexer/database github.com/cyverse-de/dataone-indexer/model"
+
+node('docker') {
+    slackJobDescription = "job '${env.JOB_NAME} [${env.BUILD_NUMBER}]' (${env.BUILD_URL})"
+
+    try {
+        checkout scm
+
+        // Get the git commit hash and descriptive version.
+        gitCommit = sh(returnStdout: true, script: "git rev-parse HEAD").trim()
+        descriptiveVersion = sh(returnStdout: true, script: "git describe --long --tags --dirty --always").trim()
+
+        // Load the service properties.
+        service = readProperties file: 'service.properties'
+
+        // Docker container names.
+        dockerTestRunner = "test-${env.BUILD_TAG}"
+        dockerPusher = "push-${env.BUILD_TAG}"
+
+        // Docker image names.
+        dockerRepo = "test-${env.BUILD_TAG}"
+        dockerPushRepo = "${service.dockerUser}/${service.repo}:${env.BRANCH_NAME}"
+
+        try {
+            // Display the git commit and descriptive version.
+            echo "git commit: ${gitCommit}"
+            echo "descriptive version: ${descriptiveVersion}"
+
+            // Build the Docker image.
+            stage('Docker Build') {
+                buildDockerImage(gitCommit, descriptiveVersion, dockerRepo)
+            }
+
+            // Test the service.
+            stage('Test') {
+                testGoService(dockerTestRunner, dockerRepo, packagesToTest)
+            }
+
+            // Push the docker image.
+            milestone 100
```

Comment:
Once again small, but it's usually good to put a `milestone 0` at or near the top of the file. The way the milestones work is that if a job reaches a milestone, it kills jobs which are older than it and *have reached a previous milestone* (emphasis mine) -- so with the milestones only here and below, it won't kill any jobs until they try to push, rather than the probably more desirable behavior of killing older jobs that are still building or running tests as soon as any newer job reaches this line. Having every job go through a milestone 0 first means that every job will have reached a previous milestone whenever you try to use one.

Reply:
Cool. Will do.